File: 234.palindrome-linked-list.py
#
# @lc app=leetcode id=234 lang=python3
#
# [234] Palindrome Linked List
#
from typing import List
# @lc code=start
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:


        values = []
        current = head
        while current != None:
            values.append(current.val)
            current = current.next

        return self.check(values)
    
    def check(self, stack: List) -> bool:
        for i in range(len(stack) // 2):
            print(stack[i], stack[-(i + 1)])
            if stack[i] != stack[-(i + 1)]:
                return False
        return True

"""
time: O(n)인데
space: O(n)임
"""

# @lc code=end

File: Linked List/src/234.palindrome-linked-list.js
/*
 * @lc app=leetcode id=234 lang=javascript
 *
 * [234] Palindrome Linked List
 */

// Definition for singly-linked list.
class ListNode {
  constructor(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
  }
}
// @lc code=start
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
  if (head.next === null) {
    return true;
  } else if (head.next.next === null) {
    if (head.val === head.next.val) {
      return true;
    }
    return false;
  }

  let slow = head;
  let fast = head;
  const slowHead = new ListNode(); // 노드를 역순으로 저장할 새 linked list
  let slowNode = slowHead;
  let temp;
  // O(1) space

  // O(n) time
  while (fast.next?.next ?? null !== null) {
    // fast는 두 칸씩 이동, slow는 한 칸씩 이동

    temp = slowNode.next;
    slowNode.next = slow;

    slow = slow.next;
    fast = fast.next.next;

    slowNode.next.next = temp;
  }
  
  // 노드가 짝수개면, slow를 한칸 더 이동시켜야 가운데로 나뉨
  if (fast.next !== null) {
    temp = slowNode.next;
    slowNode.next = slow;

    slow = slow.next;
    
    slowNode.next.next = temp;
  } else {
    slow = slow.next;
  }

  slowNode = slowHead.next;
  while (slow !== null) {
    if (slow.val !== slowNode.val) {
      return false;
    }
    slowNode = slowNode.next;
    slow = slow.next;
  }

  return true;
};
// @lc code=end

